### ISSUE TITLE
Vagrant Support

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#770]
@JustinProminic requested that we add support for [Vagrant](https://www.vagrantup.com/intro) in Moonshine.  This will let the user quickly spin up a VM for a test environment or other purpose.  Some possible features:
- run commands like `vagrant up` or `vagrant halt` to manage an existing configuration
- Edit the configuration in Vagrantfile
- Initialize an existing project with Vagrant?   I think we would want to keep Vagrant as an "addin " available for all projects rather than having a dedicated template.
- Monitor the status of the vagrant instance?

Here are some [installation instructions](https://www.vagrantup.com/docs/installation).  Vagrant also requires at least one configured hypervisor to work (I have generally used VirtualBox). 

We need to discuss this more before we start on implementation.

**Comment by JoelProminic** [comment]
More notes from @JustinProminic

>After the next release, which is focused on just getting the Mac App Store version working again plus a few other mostly done already features like the new homepage (I think?), the very next thing we need to do is add to the SDK Installer and awareness in the Getting Started page:
>
>Vagrant
>VirtualBox
>
>
>These is going to be a prerequisite for a lot of things Bing is working on, such as Structure4Notes.  
>
>Note that we do not need to INSTALL Vagrant and VirtualBox, we simply need to detect whether they are present on the Getting Started and add them to the SDK Menu and then auto-download them.   We should then just auto-launch the installer for each one, or provide a button for the user to launch the installer.
>
>
>The real point here is to add awareness of Vagrant and VirtualBox, make it as easy as possible for the user to install and launch them without having to separately go visit the websites of the vendors.
>
>
>
>I have been working on a Vagrant build of Domino 12 that will allow us to plug Moonshine Domino related software builds directly into a live, locally running and locally built, Domino 12 server that can be quickly destroyed and recreated automatically as the user wants to experiment.
>
>Bing is going to need Vagrant to have the NSF to Domino Visual Editor XML converter local to the user.
>
>Bing is going to need Vagrant to have Structure4Notes Designer, Importer, and Broker local.
>
>We need to get all of these things tied together.  They will make some wonderful videos of combining all this technology.
>
>
>Vagrant depends upon VirtualBox for the way I am building the Vagrant configs, which require shared folder sync only available in VirtualBox for bi-directional host to guest VM file synchronization for certain folders.
>
>This will allow Moonshine to build something, it will automatically get injected into the Domino or Structure4Notes Vagrant VM, and the user can instantly see what they just coded.

**Comment by JoelProminic** [comment]
I added an example Vagrant project for Royale here:  https://github.com/Moonshine-IDE/VagrantRoyaleExample

This will run the Royale project on a server created with Vagrant.  This is not a very practical example, but it is relatively fast and simple compared to the Vagrant files that we will use in general.

**Comment by rat-moonshine** [comment]
For quick reference on installation and uninstallation process, here's some earlier discussions with @JoelProminic .

### Installation

> If you aren't familiar with Vagrant, here are some brief instructions on how to get started with Vagrant.  We'll need to automate the Vagrant and VirtualBox installations as part of the Moonshine integration.
> 
> Download and install Vagrant from here: https://www.vagrantup.com/downloads.html
> 
> Download and install VirtualBox from here: https://www.virtualbox.org/wiki/Downloads
> If you get an error installing VirtualBox on High Sierra or higher, see this page: https://github.com/docksal/docksal/issues/417
> 
> Once this is done, you should be able to initialize a project like this:
> 
> cd path/to/your/project
> vagrant init
> 
> And download the desired box like this:
> 
> vagrant box add centos/7
> 
> You can find more boxes here:  https://app.vagrantup.com/boxes/search
> The Vagrantfile documentation is here:  https://www.vagrantup.com/docs/vagrantfile

### Uninstallation

> I see that the uninstall process is more complicated than usual, but it doesn't look too bad.  I haven't needed to try this myself, but I'll probably need to do this for testing:
> 
> https://www.vagrantup.com/docs/installation/uninstallation
> https://osxdaily.com/2019/01/25/uninstall-virtualbox-mac-completely/

**Comment by rat-moonshine** [comment]
Possible areas to cover:

- [ ] New "Settings" page for Vagrant path setup
- [ ] Generation of context-menu items in sidebar against particular file-name/type
- [ ] Functional implementation to Vagrant context-menu options
- [ ] New entry addition for Vagrant in MSDKI 
- [ ] Vagrant setup through MSDKI
- [ ] Detection and auto-setup of Vagrant in Moonshine-IDE through MSDKI setup logic

**Comment by rat-moonshine** [comment]
Following are the list of items suggested by @JustinProminic for Vagrant and MacPorts (#921 ) implementation:

1. Add to Getting Started detection:  Vagrant - `/usr/local/bin/vagrant`
2. Add to Getting Started detection:  MacPorts - `/opt/local/bin/port`
3. On the About screen, show the Vagrant version
4. On the About screen, show the MacPorts version
5. Add Vagrant to SDK Installer
6. Add MacPorts to the SDK Installer (for ref: https://stackoverflow.com/questions/17308629/specify-vagrantfile-path-explicity-if-not-plugin)
7. Context-menu items in sidebar - `Up`, `Halt`, `Reload (to sync files)`, `SSH`. Populate against named file `Vagrantfile`
8. Vagrant up needs to effectively do what this wrapper script does:
```
#!/bin/bash
## Simple wrapper for vagrant up to write to a log as well as the console
LOG_FILE=vagrant_up.log
vagrant up  2>&1 | tee $LOG_FILE
echo "Log written to:  $LOG_FILE"
```
9. Have a separate Console just for the Vagrant up log (ideally)
10. Project sync directory so that as code changes, it gets auto-deployed into Vagrant VM: https://github.com/DominoIDP/domino_idp/blob/master/Vagrantfile

(**Note:** For 1 and 2, usually auto-setup logic handles by MSDKI codes in Moonshine; Since it'd be better to keep the MSKDI changes in **Phase 2**, we can arrange a static path-check logic in Moonshine side for now)

**Comment by JoelProminic** [comment]
To clarify, MacPorts has nothing to do with Vagrant and VirtualBox.  See #921 for MacPorts.

My suggestion for the developmemnt order on Vagrant support is:
1. Add new Settings tab for Vagrant.  This should only require a setting for the Vagrant Path
2. Add Vagrant to About page 
3. Add context menu item:  Vagrant > vagrant up
4. Implement "vagrant up" action, writing to a log file
5. Add a new tab on the bottom to view Vagrant output
6.  Add entry in Getting Started
7. Add "vagrant reload" and "vagrant halt" actions
8. Add "vagrant ssh" action, which opens a Terminal/Command Line
9. MSDKI installation for Vagrant - this requires two large installers, and I'm not sure how well we can automate this.

We can break this into multiple issues if desired.

**Comment by rat-moonshine** [comment]
Thank you, @JoelProminic , for setting up the ordering priorities to this implementation. It'll help me.

**Comment by rat-moonshine** [comment]
Some notes on Vagrant path setting. 

I had two options to set path for the Vagrant:
- by folder, i.e. `/usr/local/bin`
- by file, i.e. `/usr/local/bin/vagrant`

I chose to go by the latter for now because of the following reasons.

The original Vagrant installation path on macOS is `/opt/vagrant`. This also consists of `bin/vagrant`. Because of this extra `bin` folder I chose to the path-setting by-file way. 

If I expects that one should go by-folder way then Moonshine would set the `VAGRANT_HOME` as `/opt/vagrant` and therefore calculate the full path as `VAGRANT_HOME`+`/bin/vagrant`. But an user also could select `/usr/local/bin` - in that case the calculated addition of `+/bin/vagrant` would be false. To overcome this I could add some static path checking or adds a message that the path should contain `bin/vagrant` - but this will only add more confusion and handlings. Thus, I chose to go by-file selection way to correct executable, like in Git or SVN.

**Comment by rat-moonshine** [comment]
While running `vagrant up` in Moonshine, console shows following error and terminates the process:

```
: vagrant up
: Missing vbguest plugin.  Install using 'vagrant plugin install vagrant-vbguest'
```

I remember last night I was required to run this particular command manually on Terminal before running `vagrant up`. Do we need this to be a standard to execute the `vagrant plugin install vagrant-vbguest` before running `vagrant up` - every time?

**Comment by rat-moonshine** [comment]
Covered so far:
- `Moonshine > Settings > Vagrant`
- Moonshine environment variable setup for Vagrant
- Context menu generation against `Vagrantfile`
- `vagrant up` is running - [check](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/770#issuecomment-1002903377) for holding
- New MSDKI entry for Vagrant and requisite initial handling
- New Vagrant entry in Getting Started
- Vagrant version tests in About 
- Other related fixes

**Comment by JoelProminic** [comment]
You shouldn't need to install vagrant-vbguest before every `vagrant up` call, but it is something we should include with the Vagrant installation in MSDKI when we get to this point.   It is odd that it showed as uninstalled for you though, so let me know if you see this problem again.

**Comment by rat-moonshine** [comment]
I'm seeing the error during every run.

**Comment by rat-moonshine** [comment]
I'm not entirely sure what is going on, but proposing an inspection to these logs and suggest me.

Since [this](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/770#issuecomment-1002903377) run error, I started to execute the command as this:
`vagrant plugin install vagrant-vbguest; vagrant up 2>&1 | tee vagrant_up.log`

**First run output:**
```
: Installing the 'vagrant-vbguest' plugin. This can take a few minutes...
: Installed the plugin 'vagrant-vbguest (0.30.0)'!
: Bringing machine 'default' up with 'virtualbox' provider...
: ==> default: Clearing any previously set forwarded ports...
: ==> default: Clearing any previously set network interfaces...
: ==> default: Preparing network interfaces based on configuration...
:     default: Adapter 1: nat
: ==> default: Forwarding ports...
:     default: 8080 (guest) => 8080 (host) (adapter 1)
:     default: 22 (guest) => 2222 (host) (adapter 1)
: ==> default: Booting VM...
: ==> default: Waiting for machine to boot. This may take a few minutes...
:     default: SSH address: 127.0.0.1:2222
:     default: SSH username: vagrant
:     default: SSH auth method: private key
:     default: Warning: Connection reset. Retrying...
:     default: Warning: Remote connection disconnect. Retrying...
: ==> default: Machine booted and ready!
: Got different reports about installed GuestAdditions version:
: Virtualbox on your host claims:   
: VBoxService inside the vm claims: 6.1.30
: Going on, assuming VBoxService is correct...
: [default] GuestAdditions seems to be installed (6.1.30) correctly, but not running.
: Got different reports about installed GuestAdditions version:
: Virtualbox on your host claims:   
: VBoxService inside the vm claims: 6.1.30
: Going on, assuming VBoxService is correct...
: Redirecting to /bin/systemctl start vboxadd.service
: Job for vboxadd.service failed because the control process exited with error code. See "systemctl status vboxadd.service" and "journalctl -xe" for details.
: Redirecting to /bin/systemctl start vboxadd-service.service
: Job for vboxadd-service.service failed because the control process exited with error code. See "systemctl status vboxadd-service.service" and "journalctl -xe" for details.
: Got different reports about installed GuestAdditions version:
: Virtualbox on your host claims:   
: VBoxService inside the vm claims: 6.1.30
: Going on, assuming VBoxService is correct...
: VirtualBox Guest Additions: Starting.
: VirtualBox Guest Additions: Building the VirtualBox Guest Additions kernel 
: modules.  This may take a while.
: VirtualBox Guest Additions: To build modules for other installed kernels, run
: VirtualBox Guest Additions:   /sbin/rcvboxadd quicksetup <version>
: VirtualBox Guest Additions: or
: VirtualBox Guest Additions:   /sbin/rcvboxadd quicksetup all
: VirtualBox Guest Additions: Kernel headers not found for target kernel 
: 3.10.0-1127.el7.x86_64. Please install them and execute
:   /sbin/rcvboxadd setup
: modprobe vboxguest failed
: The log file /var/log/vboxadd-setup.log may contain further information.
: ==> default: Checking for guest additions in VM...
:     default: No guest additions were detected on the base box for this VM! Guest
:     default: additions are required for forwarded ports, shared folders, host only
:     default: networking, and more. If SSH fails on this machine, please install
:     default: the guest additions and repackage the box to continue.
:     default: 
:     default: This is not an error message; everything may continue to work properly,
:     default: in which case you may ignore this message.
: The following SSH command responded with a non-zero exit status.
: Vagrant assumes that this means the command failed!
: /usr/sbin/rcvboxadd setup
: Stdout from the command:
: VirtualBox Guest Additions: Starting.
: VirtualBox Guest Additions: Building the VirtualBox Guest Additions kernel 
: modules.  This may take a while.
: VirtualBox Guest Additions: To build modules for other installed kernels, run
: VirtualBox Guest Additions:   /sbin/rcvboxadd quicksetup <version>
: VirtualBox Guest Additions: or
: VirtualBox Guest Additions:   /sbin/rcvboxadd quicksetup all
: VirtualBox Guest Additions: Kernel headers not found for target kernel 
: 3.10.0-1127.el7.x86_64. Please install them and execute
:   /sbin/rcvboxadd setup
: Stderr from the command:
: modprobe vboxguest failed
: The log file /var/log/vboxadd-setup.log may contain further information.
```
**Second run output (same session):**
```
: Installing the 'vagrant-vbguest' plugin. This can take a few minutes...
: Installed the plugin 'vagrant-vbguest (0.30.0)'!
: Bringing machine 'default' up with 'virtualbox' provider...
: ==> default: Machine already provisioned. Run `vagrant provision` or use the `--provision`
: ==> default: flag to force provisioning. Provisioners marked to run always will still run.
: ==> default: Running provisioner: shell...
:     default: Running: inline script
: ==> default: Running provisioner: shell...
:     default: Running: inline script
:     default: #### Server is starting. See log at /home/vagrant/server.log ####
:     default: #### Open the application at:  http://127.0.0.1:8080/index.html ####
: ==> default: Running provisioner: shell...
:     default: Running: inline script
```

While the **second run output** seems to start the Vagrant successfully, I'm wondering why the **first run** terminates abruptly.

**Comment by rat-moonshine** [comment]
Covered so far:
- Separate console window for Vagrant-only output
- Tailored context-menu

**Comment by JoelProminic** [comment]
Yeah, I noticed in my testing that the script worked on the second time even if there was an error on the shared folder configuration.  I think that is is skipping the shared folder setup step in this case and starting anyway.  Does the server work properly in your test (and is ~/server populated)?

Looking at this error, I think there is something wrong with your current VirtualBox installation.  You may want to try reinstalling:

    : Got different reports about installed GuestAdditions version:  
    : Virtualbox on your host claims:   
    : VBoxService inside the vm claims: 6.1.30

**Comment by rat-moonshine** [comment]
> Does the server work properly in your test (and is ~/server populated)?

No, running http://127.0.0.1:8080/index.html on a browser gives me this error in browser:
```
Error response
Error code 404.
Message: File not found.
Error code explanation: 404 = Nothing matches the given URI.
```
I also receive following error in running Vagrant console in Moonshine:
```
:     default: #### Server is starting. See log at /home/vagrant/server.log ####
:     default: #### Open the application at:  http://127.0.0.1:8080/index.html ####
: ==> default: Running provisioner: shell...
:     default: Running: inline script
:     default: 10.0.2.2 - - [03/Jan/2022 05:22:42] code 404, message File not found
:     default: 10.0.2.2 - - [03/Jan/2022 05:22:42] "GET /index.html HTTP/1.1" 404 -
```
I tried to re-install from _VirtualBox-6.1.30-148432-OSX.dmg_, but that doesn't makes any difference from earlier tests.
![image](https://user-images.githubusercontent.com/24370870/147901789-3d310c79-f425-4cc5-b702-07a7ef90728a.png)

**Comment by rat-moonshine** [comment]
- `vagrant halt` implemented
- `vagrant reload` implemented

> Add "vagrant ssh" action, which opens a Terminal/Command Line
I'm unclear about the expected behaviour from `vagrant ssh`, I probably need more information.

**Comment by JoelProminic** [comment]
Up, Halt, and Reload work, but I don't think it is desirable to install vbguest on every execution.  We will need to do more debugging to determine why it is not working properly on your machine.  Please try:
1. Open a new Terminal window and `cd` to your project, so that we can clear the history
2. Run `vagrant destroy` to delete the instance and start fresh
3. Run `git pull` to make sure you are on the latest version
4. Run `vagrant plugin list` to confirm that the vbguest plugin is installed globally.
5. Run `vagrant up`
6. Send me the full output by email if you get an error.  

Example of `vagrant plugin list` 

    % vagrant plugin list
    vagrant-disksize (0.1.3, global)
      - Version Constraint: > 0
    vagrant-reload (0.0.1, global)
      - Version Constraint: > 0
    vagrant-scp (0.1.0, global)
      - Version Constraint: > 0
    vagrant-vbguest (0.30.0, global)

**Comment by JoelProminic** [comment]
To get familiar with `vagrant ssh`, try:
1. Start Vagrant with Vagrant > Up
2. On project, Right-click > Copy Path
3. Open a Terminal window
4. Run `cd %paste-path%`
5. Run `vagrant ssh`
6. Now you are SSHed into the Linux VM created for this project.  You can run commands like `ls server` to see the server files, or `ps -ef | grep python` to see the server process.
7. Run `exit` to leave the SSH session.

Example: 

    % vagrant ssh
    [vagrant@localhost ~]$ ps -ef | grep python
    root       652     1  0 22:32 ?        00:00:00 /usr/bin/python2 -Es /usr/sbin/tuned -l -P
    vagrant   2575  2574  0 22:33 ?        00:00:00 python -m SimpleHTTPServer 8080
    vagrant   2606  2584  0 22:43 pts/0    00:00:00 grep --color=auto python
    [vagrant@localhost ~]$ ls server
    index.html      MainContent.js.map  VagrantRoyaleExample.css     view
    library         org                 VagrantRoyaleExample.js
    MainContent.js  resources           VagrantRoyaleExample.js.map
    [vagrant@localhost ~]$ 
    [vagrant@localhost ~]$ exit
    logout
    Connection to 127.0.0.1 closed.
    % 

For the initial implementation of Vagrant > Up, we need to automate steps 2-5.  For now, I want to see this open in an external Terminal or Command Prompt so that we don't need to worry about an interactive prompt in Moonshine.

**Comment by rat-moonshine** [comment]
> Up, Halt, and Reload work, but I don't think it is desirable to install vbguest on every execution. We will need to do more debugging to determine why it is not working properly on your machine. 

While running on Terminal I see the `vbguest` installed properly:
```
% vagrant plugin list
vagrant-vbguest (0.30.0, global)
```
Running `vagrant up` therefore started the process normally without throwing an error.

I'm wondering if `vbguest` not seeing as installed-one while running from Moonshine-IDE, because the application can't see system environment variable settings.

**Comment by rat-moonshine** [comment]
> For the initial implementation of Vagrant > Up, we need to automate steps 2-5. For now, I want to see this open in an external Terminal or Command Prompt so that we don't need to worry about an interactive prompt in Moonshine.

At Moonshine side, I unable to determine when a `vagrant up` process completes so can run `vagrant ssh`. I can check the output from `vagrant up` to determine when `default: Running: inline script` prints - unfortunately, this seems to be printed many a time, so can't really sure if `up` process completed correctly or not.

Opening a separate Terminal window running scripts on it - this anyway possible in non-sandbox but might not in sandbox unless proper entitlement. I need to check.

**Comment by rat-moonshine** [comment]
I confirm executing AppleScript to open Terminal with command-running works on non-sandbox but sandbox build. For that we probably needs AppleScript execution entitlement which Apple reviewer won't give. 

Perhaps, we can think of not availing `ssh` command on sandbox build.

**Comment by rat-moonshine** [comment]
A "Configure Vagrant" option being added to context-menu in-case Vagrant is not available:

![image](https://user-images.githubusercontent.com/24370870/148194901-7b322303-0c1a-45e7-8a12-cbc1506233b7.png)

**Comment by rat-moonshine** [comment]
`vagrant ssh` added. Currently this works only on non-sandbox version. 

[Task breakups](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/770#issuecomment-1002393359)
(NOTE: All the implements and tests were done only on/for macOS, until here)

**Comment by rat-moonshine** [comment]
I have updated `vagrant ssh` to open command-prompt on Windows. 

I did other couple of tests on Windows as well. 

I see during executing `vagrant` from Moonshine - it throws following error on console:
```
: Vagrant failed to initialize at a very early stage:
: 
: The home directory you specified is not accessible. The home
: directory that Vagrant uses must be both readable and writable.
: 
: You specified: C:/HashiCorp/Vagrant/bin
```
I remember same error I had seen Vagrant path setup to `/opt/vagrant/bin`, instead of `/usr/local/bin` on macOS. I overcome the error by strictly pointing to `/usr/local/bin` (which holds the symlink file of Vagrant) on macOS.

**Comment by MarkProminic** [comment]
On Windows, my Vagrant Home Directory is defined as:

C:\Users\Markg.MVCS\.vagrant.d

This may help you with redefining the VAGRANT_HOME directory:

https://harvsworld.com/2014/change-vagrant_home-directory-windows/


You can change this to a path of your choosing, just make sure that you have read and write access to the folder.

**Comment by JoelProminic** [comment]
Vagrant > SSH is working great on macOS.  I do get an intimidating permissions warning, but this makes sense for what we are trying to do.
![image](https://user-images.githubusercontent.com/24532731/148296838-a5a8e186-e447-431b-843c-8bf1a99bcabb.png)

Justin requested that we set a title for the Terminal window.  We should be able to do this with Applescript [(example)](https://stackoverflow.com/questions/37631769/rename-window-title-in-os-x-with-applescript-or-terminal)

**Comment by JoelProminic** [comment]
I have also noticed that if I try to open a console to the Vagrant VM from the VirtualBox application, I get a permission request like this.  We think this is because the VM inherited the permissions from Moonshine-IDE.  I found it wasn't necessary to allow this permission, but it may cause problems later.

![image](https://user-images.githubusercontent.com/24532731/148296895-0a844566-562c-4d90-9f36-c327f8963384.png)

**Comment by JoelProminic** [comment]
Some other updates:

Add a "Vagrant > Destroy" action which runs ths `vagrant destroy` command.  This needs to have a confirmation prompt before running the destroy command, because this could waste a lot of time if it is clicked accidentally.

Currently, I can't open Vagrantfile in the Moonshine editor.  We think this is caused because there is no extension.  If we want to make this a more generic solution, we can create a separate issue.

How difficult would it be to add a stop/kill button in the bottom right for the Vagrant actions?  This would be similar to the red stop button for the Build & Run actions.  If this is a big change, we could move this to a separate issue.

**Comment by rat-moonshine** [comment]
> Justin requested that we set a title for the Terminal window. We should be able to do this with Applescript (example)

I was previously running the `osascript` as a single liner with script execution. This now replaced with a full-AS file generation and execution. That fixes the title problem.

**Comment by rat-moonshine** [comment]
A new `Vagrant > Destroy` option being added to context menu.

This suppose to open a confirmation popup to the user:
![image](https://user-images.githubusercontent.com/24370870/148364307-8540802c-2241-458d-8909-475d6f3ffbb0.png)

(Let me know the corrected texts, if any)

While calling `vagrant destroy` from Moonshine-IDE, following errors thrown to its console:
```
: Vagrant is attempting to interface with the UI in a way that requires
: a TTY. Most actions in Vagrant that require a TTY have configuration
: switches to disable this requirement. Please do that or run Vagrant
: with TTY.
```
Reading through some discussions (some links are as follows) it sounds like overcoming TTY problem can be doable by executing something like this:

`export VAGRANT_DETECTED_OS="$(uname)"`
(`uname` can collect by running `uname -s`. For me it returns `Darwin`)

Discussion links:
https://github.com/hashicorp/vagrant/issues/6026
https://stackoverflow.com/questions/23633276/vagrant-is-attempting-to-interface-with-the-ui-in-a-way-that-requires-a-tty

I tried to set Moonshine environment variables with `VAGRANT_DETECTED_OS=Darwin` - but that didn't fix the TTY problem during `vagrant destroy`.

Then I noticed this following command:

`vagrant destroy %vboxName% -f`
(`-f` to skip the confirmation prompt during the execution)

This worked for me. 

We can also run this as `vagrant destroy default -f`, but the `%vboxName%` value is not available to me and I'm not sure about the standard - whether it always suppose to be `default` even when there are multiple `Vagrantfile`.

**Comment by rat-moonshine** [comment]
> How difficult would it be to add a stop/kill button in the bottom right for the Vagrant actions? This would be similar to the red stop button for the Build & Run actions. If this is a big change, we could move this to a separate issue.

The stop button is not avails during `vagrant up`. 

Doing a stop will execute following actions:
- Stop the current running process
- Execute `vagrant halt` (I was unsure whether this should a followed-by execution since Stop button, let me know if we don't want to execute this)

**Comment by rat-moonshine** [comment]
> Currently, I can't open Vagrantfile in the Moonshine editor. We think this is caused because there is no extension. If we want to make this a more generic solution, we can create a separate issue.

You should able to open any textual file without an extension, now.

**Comment by rat-moonshine** [comment]
> On Windows, my Vagrant Home Directory is defined as:
> 
> C:\Users\Markg.MVCS\.vagrant.d
> 
> This may help you with redefining the VAGRANT_HOME directory:
> 
> https://harvsworld.com/2014/change-vagrant_home-directory-windows/
> 
> You can change this to a path of your choosing, just make sure that you have read and write access to the folder.

Thank you, @MarkProminic . This differs my idea of `SDK_HOME=% sdkExecutableLocation%` as we used to see.

Since Moonshine not able to access system environment variables, we set its own environment variables during its runtime, and like other SDKs - we set Vagrant's as `VAGRANT_HOME=C:\HashiCorp\Vagrant\bin`. But it sounds like Vagrant does read/write stuff to its `VAGRANT_HOME` location (?)

@JoelProminic , I changed this now as `VAGRANT_HOME=File.applicationStorageDirectory.nativePath` - this should has full access for any read/write, during a session. I'm not sure if we want this change to vagrantfile-directory instead, then I'll need to set the `export` command separately during `Vagrantfile` actions. 

I tested all the commands on macOS works smoothly, after these changes. I also confirm the Windows' [error](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/770#issuecomment-1005601673) is not triggering anymore.

**Comment by rat-moonshine** [comment]
On Windows, during `vagrant up` I see following installation error - but hopefully this not any setup error with Moonshine:

```
: Installing the 'vagrant-vbguest' plugin. This can take a few minutes...
: Installed the plugin 'vagrant-vbguest (0.30.0)'!
: There was an error while executing `VBoxManage`, a CLI used by Vagrant
: for controlling VirtualBox. The command and stderr is shown below.
: 
: Command: ["startvm", "5d3--censored--7c4", "--type", "headless"]
: 
: Stderr: VBoxManage.exe: error: Not in a hypervisor partition (HVP=0) (VERR_NEM_NOT_AVAILABLE).
: VBoxManage.exe: error: AMD-V is disabled in the BIOS (or by the host OS) (VERR_SVM_DISABLED)
: VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component ConsoleWrap, interface IConsole
```

**Comment by JoelProminic** [comment]
Vagrant > Destroy works on macOS.  Here is an updated message.  I added some clarification about what would happen with this action.

>### Confirm
>Are you sure you want to destroy the Vagrant instance for %project-path%?
>
>The virtual machine will be permanently destroyed, and will need to be recreated for future tests.
>
>The Vagrant box will \*not\* be destroyed.

**Comment by JoelProminic** [comment]
Yeah, I see that VAGRANT_HOME has a different meaning than we are used to:  https://www.vagrantup.com/docs/other/environmental-variables

> VAGRANT_HOME can be set to change the directory where Vagrant stores global state. By default, this is set to ~/.vagrant.d. The Vagrant home directory is where things such as boxes are stored, so it can actually become quite large on disk.

So, if we set this to be `File.applicationStorageDirectory.nativePath`, then:
- Boxes will be installed to the hidden application directory.  This could be a lot of the space that the user wouldn't be able to clean up easily.
- Any plugins the user installed manually would not be visible.  If the user installs the plugins from the script like @MarkProminic showed, then this only wastes initial startup time.

There are some use cases where a user might want to change VAGRANT_HOME. For example, I changed this on my old workstation to move the boxes to a separate, larger disk.  Since the macOS applications ignore the enviroment variables in ~/.profile, any custom settigns the user has defined will be ignored by default.  

We could add a Vagrant Home option in the Moonshine Vagrant settings.  The value would need to be blank by default, and VAGRANT_HOME should not be written if it is blank so that it can use the system default.  However, this ignores the other environment variables from my link above.

Another solution would be to add an interface to let the user define arbitrary environment variables, but this should be a separate issue if we choose to implement this.

For now, don't set VAGRANT_HOME.  We can create a separate issue if we want to use one of my above suggestions to set a custom VAGRANT_HOME.

**Comment by JoelProminic** [comment]
@rat-moonshine confirmed that he gets the "hypervisor partition" error when he runs the "vagrant up" command directly on Windows.  It seems his workstation does not have VM support.  @MarkProminic, we will need more help testing this once @rat-moonshine does some more cleanup.

**Comment by JoelProminic** [comment]
For the stop/kill button my intention was for this to kill the current action, rather than stopping the VM.  Some use cases:
- For the VagrantRoyaleProject, `vagrant up` does not terminate because that would kill the test server (this should be fixed in the example project, but I'll probably need to setup a service).
- In other cases, the user may realize that they ran the ran command, or see a fatal error in "vagrant up" that they need to fix.

For now, remove "vagrant halt" from the stop/kill button.  We may eventually add a status monitor for the Vagrant instance, but this should be a separate issue.

**Comment by JoelProminic** [comment]
Other tests:
- I confirmed that the Terminal window title is set to the project name now.  This should be good for now.
- I confirmed that I can open Vagrantfile in an editor now.  There is no syntax highlighting for now, but that is expected.  This is a different language (ruby), so we'll need some updates to handle the format if we want to enable syntax highlighting later.

**Comment by JoelProminic** [comment]
Also, I would like to remove the vagrant-vbguest install command from the Vagrant > Up action.  This should no longer be needed once we are using the default VAGRANT_HOME.

I have [some other options](https://github.com/Moonshine-IDE/VagrantRoyaleExample/issues/2) to automatically install this for the demo projects.

**Comment by rat-moonshine** [comment]
![image](https://user-images.githubusercontent.com/24370870/148501202-95fba89a-37fd-4da3-ab58-093601fd4de4.png)

I just noticed from last day's change - opening a non-extension file left a null value to its name in "Recent Files" section. This now fixed.

**Comment by rat-moonshine** [comment]
Updates:
- `destroy` alerts updated
- Custom set of `VAGRANT_HOME` already being removed last night. @JoelProminic notice you may see the vagrant process/output became faster than previous! I'm not sure if removing the custom `VAGRANT_HOME` setup affects to this some way 
- `vagrant halt` removed during "Stop" button click in status-bar
- `vagrant-vbguest` install removed during `vagrant up` - I confirm I don't see any problem after this change
- Vagrant logo updated in Getting Started

**Comment by JoelProminic** [comment]
I do not see the red stop button for Vagrant > Halt and Vagrant > Destroy.  Was this intentional for Vagration > Halt, so that it did not get stuck in a loop?

The updated Vagrant > Destroy message looks fine.  I'm not 100% satisfied with the text, but this is good enough for now.

I got some strange behavior on macOS for VAGRANT_HOME and vagrant-vbguest in my tests today:
1. From Terminal, run `vagrant uninstall vagrant-vbguest`
2. From Moonshine, run `vagrant up`.
3. `vagrant up` runs without reporting that the guest additions are missing

However, I see the same behavior on the Terminal, I'm going to try restarting my machine later to see if that helps.

It occurred to me that if we use the default config directory (`~/.vagrant_d`), then this could change if Moonshine uses a different home directory.  I did some quick test with this, and it looks like non-Sandbox Moonshine Development uses the normal home directory, but I remember than the AppStore build uses a directory under ~/Library/Containers.  We'll need to address this if we want to get Vagrant working for the App Store build.

**Comment by rat-moonshine** [comment]
> I do not see the red stop button for Vagrant > Halt and Vagrant > Destroy. Was this intentional for Vagration > Halt, so that it did not get stuck in a loop?

Do we need have a halt button against them? I haven't included the option because I thought `halt` or `destroy` is some sort of execute and forgot type of command, i.e. once starts can't be stopped. Let me know if we want to include the option anyway.

**Comment by rat-moonshine** [comment]
I have added the stop button against `vagrant destroy`, but this do not guarantee that it breaks the execution or terminate once started running process in shell. Rather, this stops the running process at Moonshine side. 

We can test how this affects, anyway.

**Comment by JoelProminic** [comment]
From my initial test, it seems that cancelling Vagrant > Destroy is pretty pointless - the action is fast enough that it is already too late.  I think it is nice to give the chance, though.

I'm still seeing strange behavior where my script detects that the vagrant-vbguest plugin is installed even though it does not appear in `vagrant plugin list`.   I'm wondering if there is some other location where plugins can be installed.

**Comment by JoelProminic** [comment]
We discussed the Vagrant > Destroy confirmation message in the meeting today, and we want to make some changes to the format:
- Change "Confirm!" to "Warning!"
- Move project path to new line
- "The Vagrant template will *not* be removed."

>### Confirm
>Are you sure you want to destroy the Vagrant instance for:
>
>%project-path%?
>
>The virtual machine will be permanently destroyed, and will need to be recreated for future tests.
>
>The Vagrant template will \*not\* be removed.

**Comment by rat-moonshine** [comment]
I have updated the alert text. This issue is now also merged into `master`. You can starts tests using the regular Development build.

**Comment by JoelProminic** [comment]
Revisiting the Vagrant settings, I think we should change "Vagrant Home" to "Vagrant Binary", and store the full path to `vagrant` instead.   I don't think configuring the "home" directory instead has any extra value here, unless we need it for bookmarks in the AppStore build.  We can discuss this if you feel that we should keep using the directory instead.

**Comment by rat-moonshine** [comment]
> We can discuss this if you feel that we should keep using the directory instead.

I thought about this some time during this development, and I chose to go with "home" way instead of "binary" because of following reasons:
1. It matches the other SDK location management ways (except Git and SVN, but now I think it probably possible to change to "home" way, as well)
2. It helps using already created APIs where validate a SDK by "home" way and certain validation rules, rather than writing SDK-specific ways as we have for Git/SVN

I'd still recommend to keep this in "home" way and let Moonshine validate if the location is valid.

(I feels an imbalance in the way we sets SDKs in Settings where most of the SDKs set by "home" way and Git/SVN set by "binary" way. It also felt as we've no work similarities in the IDE)

**Comment by rat-moonshine** [comment]
- Updated alert texts on 'vagrant destroy' 
- Changed Vagrant context-menu ordering
- Fixed auto-update in Settings upon auto-detection from Getting Started for Vagrant and MacPorts

**Comment by Aszusz** [comment]
I'm testing Vagrant support on Windows using https://github.com/Moonshine-IDE/VagrantRoyaleExample/
There was a couple of UX issues that I will describe later but for now, I cannot get past this error message:

```
: "C:\HashiCorp\Vagrant\bin\vagrant.exe" up > vagrant_up.log | type vagrant_up.log 
: Log file location: C:\data\repos\prominic\VagrantRoyaleExample.git\vagrant_up.log 
: The following SSH command responded with a non-zero exit status.
: Vagrant assumes that this means the command failed!
: 
: umount /mnt
: 
: Stdout from the command:
: 
: 
: 
: Stderr from the command:
: 
: umount: /mnt: not mounted
```

**Comment by Aszusz** [comment]
On the other hand, on **MacOS** I'm getting this error:

```
: "/usr/local/bin/vagrant" up 2>&1 | tee vagrant_up.log 
: Log file location: /Users/aszusz/Repos/VagrantRoyaleExample/vagrant_up.log 
: The provider 'virtualbox' that was requested to back the machine
: 'default' is reporting that it isn't usable on this system. The
: reason is shown below:
: Vagrant could not detect VirtualBox! Make sure VirtualBox is properly installed.
: Vagrant uses the `VBoxManage` binary that ships with VirtualBox, and requires
: this to be available on the PATH. If VirtualBox is installed, please find the
: `VBoxManage` binary and add it to the PATH environmental variable.
```

I can try to fix PATH variable but I remember vagrant up working correctly not so long ago and I don't know what changed.

**Comment by JoelProminic** [comment]
Thanks for testing @Aszusz!

The Windows error looks like the error we see when the vagrant-vbguest plugin is not installed.  For example:  https://github.com/dotless-de/vagrant-vbguest/issues/399

However, I have [a check for this](https://github.com/Moonshine-IDE/VagrantRoyaleExample/blob/b247263020a1cdeb8a9376eb19bbb2527b52a5b7/Vagrantfile#L81) in the test Vagrant project, so I don't understand why it is not triggering an error in this case.  You could try reinstalling the plugin as described in the link above.

For the macOS installer, it looks like VirtualBox is not installed properly for some reason.  We can review this in the meeting.

**Comment by rat-moonshine** [comment]
I saw a different error while `vagrant up` running on Windows, however this probably on different reason than @Aszusz 's:

```
: "C:\HashiCorp\Vagrant\bin\vagrant.exe" up > vagrant_up.log | type vagrant_up.log 
: Log file location: F:\DevareaLocal\Prominic\GIT\VagrantRoyaleExample\vagrant_up.log 
: There was an error while executing `VBoxManage`, a CLI used by Vagrant
: for controlling VirtualBox. The command and stderr is shown below.
: 
: Command: ["startvm", "5d3fc0a0-censored-3767c4", "--type", "headless"]
: 
: Stderr: VBoxManage.exe: error: Not in a hypervisor partition (HVP=0) (VERR_NEM_NOT_AVAILABLE).
: VBoxManage.exe: error: AMD-V is disabled in the BIOS (or by the host OS) (VERR_SVM_DISABLED)
: VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component ConsoleWrap, interface IConsole
```

**Comment by Aszusz** [comment]
@rat-moonshine I had that one too. You need to enable virtualization in the BIOS. Do you have Intel or AMD processor? This is one of the issues I want to review at the meeting.

**Comment by Aszusz** [comment]
The motherboards for AMD processors tend to hide that setting pretty deep. For my board (MSI B450-A PRO MAX) I found those steps:
- Boot into Bios using the Del key
- Select "Overclocking"
- Go to "OC Explore Mode" and turn that to "Expert"
- Scroll down to "CPU Features"
- Select "SVM Mode" and "Enable"
- Save end exit

**Comment by rat-moonshine** [comment]
Thanks @Aszusz . That seems not very painful to search in my case once I knew what I was looking for ("SVM Mode"). 

I confirm that clears the 'hypervisor partition' error for me.

**Comment by rat-moonshine** [comment]
I see some blocker (for now) while I was doing more tests on Windows after able to overcome the 'hypervisor partition' [error](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/770#issuecomment-1016679743).

One of the requirement was to save `vagrant` command output to a _log_ file while makes its `stdOut` to console. This was written in following way between the OS:
```
if (ConstantsCoreVO.IS_MACOS)
{
	command = '"'+ vagrantBinPath +'" up 2>&1 | tee vagrant_up.log';
}
else
{
	command = '"'+ vagrantBinPath +'" up > vagrant_up.log | type vagrant_up.log';
}
```
The problem I noticed while running on Windows is that the output never available to `stdOut` until the process completes or terminates with the above script. Unfortunately, that much I able to find while searching through different articles to use `type` on command-line for both file-save and print-output.

Vagrant process is time consuming and until the process exits, console do not prints anything, because of the above limitation, unlike `tee` usage on macOS.

I see we have a few options to overcome this problem on Windows:
1. Provide new shell-command which effectively `stdOut` while running and also saves to a file
2. Consider to remove save to file feature on Windows
3. Use PowerShell instead of CMD

I tried to do some investigation to use PowerShell instead of CMD to run its `NativeProcess`. 

On Windows, Moonshine generates a batch file along with the `set` environment-commands during a `NativeProcess` run, so it can avail all the Moonshine's environment-variable during that particular shell-run. Moonshine runs that generated batch file when is ready.

While running the `NativeProcess` with PowerShell, the console output the following:
```
: "C:\HashiCorp\Vagrant\bin\vagrant.exe" up 2>&1 | tee vagrant_up.log 
: Log file location: F:\DevareaLocal\Prominic\GIT\VagrantRoyaleExample\vagrant_up.log 
: 'tee' is not recognized as an internal or external command,
: operable program or batch file.
```
This as like it is try to run the batch file using CMD. However, I confirm the `NativeProcess` was executing with PowerShell.

Strangely, while I tried to run the batch file on a standalone PowerShell prompt, I faced same error as if it's trying to run the batch file with CMD only:

```
// attempt 1
> cmd.exe /c 'C:\Users\Devsena\.....\8DE2774C-ED0D-C9C0-20EC-767C11E2AB53.cmd'
'tee' is not recognized as an internal or external command,
operable program or batch file.

// attempt 2
> & 'C:\Users\Devsena\..........\8DE2774C-ED0D-C9C0-20EC-767C11E2AB53.cmd'
'tee' is not recognized as an internal or external command,
operable program or batch file.

// attempt 3
> $A = Start-Process -FilePath 'C:\Users\Devsena\.......\8DE2774C-ED0D-C9C0-20EC-767C11E2AB53.cmd' -Wait -NoNewWindow
'tee' is not recognized as an internal or external command,
operable program or batch file.
```
@joshtynjala , you have [worked](https://github.com/feathersui/feathersui-starling-sdk-manager/blob/3a5fcd7015f102e66a2ab10f1bc21a0e67485e71/ant_on_air/src/org/apache/flex/ant/tags/Unzip.as#L143) using PowerShell before while working on ant-on-air, do you have an idea why it is above (?)

**Comment by Aszusz** [comment]
@rat-moonshine I think your PowerShell process starts another CMD process under the hood. Maybe try changing script extension from `.cmd` to `.ps` and see what happens when you Start-Process it.

**Comment by rat-moonshine** [comment]
Thanks @Aszusz . Unfortunately, that didn't ran well:

```
> Start-Process -FilePath 'C:\Users\Devsena\.......\A79DBCE9-4D0C-5D17-058B-781C68D88B14.ps' -Wait -NoNewWindow
Start-Process : This command cannot be run due to the error: %1 is not a valid Win32 application.
At line:1 char:1
+ Start-Process -FilePath 'C:\Users\Devsena\.....\moonshine ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Start-Process], InvalidOperationException
    + FullyQualifiedErrorId : InvalidOperationException,Microsoft.PowerShell.Commands.StartProcessCommand
```
While running `> & 'C:\Users\Devsena\........\A79DBCE9-4D0C-5D17-058B-781C68D88B14.ps` opens me this:
![image](https://user-images.githubusercontent.com/24370870/150370307-046e9568-712e-4a6a-88a8-7cc273512aaf.png)

For a reference, here is the generated batch file:
```
@echo off
set FLEX_HOME=C:\MoonshineSDKs\Flex_SDK\Flex_4.16.1_AIR_32.0
set AIR_SDK_HOME=C:\MoonshineSDKs\Flex_SDK\Flex_4.16.1_AIR_32.0
set JAVA_HOME=C:\MoonshineSDKs\Java\openjdk-11.0.10
set ANT_HOME=C:\MoonshineSDKs\Ant\apache-ant-1.10.1
set MAVEN_HOME=C:\MoonshineSDKs\Maven\apache-maven-3.6.0
set GRAILS_HOME=C:\MoonshineSDKs\Grails\grails-3.3.10
set HAXE_HOME=C:\HaxeToolkit\haxe
set NEKO_HOME=C:\HaxeToolkit\neko
set GIT_HOME=C:\MoonshineSDKs\Git\git-2.30.1
set PATH=%FLEX_HOME%\bin;%AIR_SDK_HOME%\bin;%JAVA_HOME%\bin;%ANT_HOME%\bin;%MAVEN_HOME%\bin;%GRAILS_HOME%\bin;%HAXE_HOME%;%NEKO_HOME%;%PATH%
"%GIT_HOME%\bin\git" config --global http.sslCAInfo "%GIT_HOME%\mingw64\ssl\cert.pem"
"C:\HashiCorp\Vagrant\bin\vagrant.exe" up 2>&1 | tee vagrant_up.log
```

**Comment by Aszusz** [comment]
@rat-moonshine 
Oh, yeah, sorry. You can go back to `.cmd` but you need to use PowerShell in the same pipe you are using tee.
Try using this as the last line of the script:
```
powershell ""C:\HashiCorp\Vagrant\bin\vagrant.exe" up 2>&1 | tee vagrant_up.log"
```

**Comment by rat-moonshine** [comment]
Thanks @Aszusz . That might work. I try adding this at the end of the batch file and run on CMD:
`"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" "C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log"`

This seems to work on standalone command-line. I need to test within Moonshine.

**Comment by Aszusz** [comment]
And by the way, if you feel like using PowerShell may be of benefit to you, I can help you rework cmd scripts to ps scripts.

**Comment by rat-moonshine** [comment]
Thank you @Aszusz . That would be helpful. At this moment, we're not ready to replace existing CMD ways with PowerShell on Windows. It'd be a big change. Currently, I'm looking for a way to resolve the current problem with Vagrant, and hopefully your suggestion will be fruitful )

**Comment by JoelProminic** [comment]
> Thanks @Aszusz . That might work. I try adding this at the end of the batch file and run on CMD: `"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" "C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log"`
> 
> This seems to work on standalone command-line. I need to test within Moonshine.

This command works for me in a bash script.  Note the "2>&1" - this was missing from my test commands today, and it is missing for the current Moonshine command.

Do we need to use the full path for powershell?  I saw some discussion for the default paths [here](https://stackoverflow.com/questions/4145232/path-to-powershell-exe-v-2-0).

    64 bits version: C:\Windows\System32\WindowsPowerShell\v1.0\
    32 bits version: C:\Windows\SysWOW64\WindowsPowerShell\v1.0\
    
    We could use this if we want to handle the case where %SystemRoot% is at a different location:
    
    %SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe

**Comment by Aszusz** [comment]
> Do we need to use the full path for powershell?

I'd assume PowerShell should already be in PATH but best to check with @MarkProminic 
Also do we need to worry about ExecutionPolicy?

**Comment by JoelProminic** [comment]
Hmm, [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.security/set-executionpolicy?view=powershell-7.2) could be a problem if you are running it from a script, since your script won't be signed.  If you see problems with this, I would recommend trying to run this command without a script - hopefully it can work as a one-liner.

**Comment by JoelProminic** [comment]
I met with @MarkProminic, @Aszusz, and Jesse today to do some testing on macOS and Windows

### macOS

We tried to test this on Jesse's mac, but we found that her hardware did not support VirtualBox.  

We were also able to reproduce the issue with missing VirtualBox.   After some investigation, we wound that the install location for VirtualBox (and Vagrant), `/usr/local/bin`, is not included in the default PATH on macOS.   It seems that we will need to configure "VirtualBox Home", and add it to the PATH.   We were able to test this by adding a symlink to VBoxManage in one of the existing MoonshineSDKs directories.

Also note that VirtualBox will require macOS security permissions in order to run properly.  The user should normally be prompted automatically for this.


### Windows

The `tee` command discussed above will be important to monitor the status of the Vagrant task in Moonshine.  This was a major annoyance when testing today, so I'd like to get this resolve soon.

`Vagrant > SSH` did not work.  It opened a Command Line to the correct directory, but the `vagrant ssh` command didn't run.

We also found some errors with the Vagrantfile for VagrantRoyaleExample on Windows.  I'll create an new issue for this on that repository.

**Comment by JoelProminic** [comment]
Some more details for "VirtualBox Home":
- This should be added in the Vagrant settings page
- Title:  "(Optional) VirtualBox Home - Parent directory of VBoxManage
- This should be automatically detected based on the VBoxManage command.  The default path for MacOS is `/usr/local/bin/VBoxManage`
- If the value is populated, then define a `VIRTUALBOX_HOME` environment variable and add `$VIRTUALBOX_HOME` entry in the path
- Do not send any alerts if VIRTUALBOX_HOME is not populated, so that the user has the option to use a different provider if it is available.

**Comment by JoelProminic** [comment]
When revisiting my tests on macOS, I see that I am also now getting the VirtualBox path error:

    : "/usr/local/bin/vagrant" halt 
    : The provider 'virtualbox' that was requested to back the machine
    : 'default' is reporting that it isn't usable on this system. The
    : reason is shown below:
    : Vagrant could not detect VirtualBox! Make sure VirtualBox is properly installed.
    : Vagrant uses the `VBoxManage` binary that ships with VirtualBox, and requires
    : this to be available on the PATH. If VirtualBox is installed, please find the
    : `VBoxManage` binary and add it to the PATH environmental variable.

It seems that `/usr/local/bin` was not on `PATH` for Catalina, either!  I suspect that `/usr/local/bin` was on the PATH previously for a different SDK, like Haxe/Neko and VAGRANT_HOME (now disabled).

I checked that `VIRTUALBOX_HOME` does not have special meaning for VirtualBox, so it should be fine to use it.

**Comment by JoelProminic** [comment]
I confirmed that changing the Haxe and Neko Home to /usr/local/bin cleared the VBoxManage error in Moonshine.   This is not a proper solution, though - it is just a temporary workaround.

**Comment by JoelProminic** [comment]
I tried to open vagrant_up.log in Moonshine, and I got this prompt:

![image](https://user-images.githubusercontent.com/24532731/150585223-902ddbf6-1bea-4871-a6e4-ee84a6c0f84b.png)

Moonshine should treat .log files as .txt files.   Is this checking something besides the extension?

**Comment by rat-moonshine** [comment]
I have added a new "VirtualBox Home" under `Settings > Vagrant`.

![image](https://user-images.githubusercontent.com/24370870/150731753-58bdd6f9-70f8-4c94-bc21-46b28e35475c.png)

Some notes, 
- The new setting is provided through Moonshine only and there is not separate reference added through MSDKI or the Getting Started
- The default location set to `/usr/local/bin` on macOS and `$programFiles/Oracle/VirtualBox` on Windows. The path on macOS will auto-setup only on non-sandbox build (as was for Vagrant) since the directory will explicitly need bookmarking on sandbox 
- A `VIRTUALBOX_HOME` variable sets to environment setting based on the validity of the path

Please, check.

**Comment by rat-moonshine** [comment]
On executing PowerShell through batch-command on Windows, I have adjusted Vagrant run-codes to accommodate PowerShell path to run with `tee`:
- 32/64 bit path detection 
- Running with `tee` instead of earlier `type` command - I confirm the outputs are available immediately to Moonshine console
- Shall notify in console if PowerShell is not available on any particular Windows

However, I see other problem during `vagrant up` run on Windows - but this could be a separate thing:

```
: "C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe" "C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log" 
: Log file location: F:\DevareaLocal\Prominic\GIT\VagrantRoyaleExample\vagrant_up.log 
: Bringing machine 'default' up with 'virtualbox' provider...
: ==> default: Clearing any previously set forwarded ports...
: ==> default: Clearing any previously set network interfaces...
: ==> default: Preparing network interfaces based on configuration...
:     default: Adapter 1: nat
: ==> default: Forwarding ports...
:     default: 8080 (guest) => 8086 (host) (adapter 1)
:     default: 22 (guest) => 2222 (host) (adapter 1)
: ==> default: Booting VM...
: C:\HashiCorp\Vagrant\bin\vagrant.exe : There was an error while executing `VBoxManage`, a CLI used by Vagrant
: At line:1 char:1
: + C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log
: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:     + CategoryInfo          : NotSpecified: (There was an er...used by Vagrant:String) [], RemoteException
:     + FullyQualifiedErrorId : NativeCommandError
:  
: for controlling VirtualBox. The command and stderr is shown below.
: Command: ["startvm", "5d3fc0a0-90d6-43e7-97ba-5b6f273767c4", "--type", "headless"]
: Stderr: VBoxManage.exe: error: The virtual machine 'VagrantRoyaleExample_default_1641467976898_91276' has terminated 
: unexpectedly during startup with exit code 1 (0x1).  More details may be available in 'C:\Users\Shadow\VirtualBox 
: VMs\VagrantRoyaleExample_default_1641467976898_91276\Logs\VBoxHardening.log'
: VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component MachineWrap, interface IMachine
```

**Comment by rat-moonshine** [comment]
On executing PowerShell through batch-command on Windows, I have adjusted Vagrant run-codes to accommodate PowerShell path to run with `tee`:
- 32/64 bit path detection 
- Running with `tee` instead of earlier `type` command - I confirm the outputs are available immediately to Moonshine console
- Shall notify in console if PowerShell is not available on any particular Windows

However, I see other problem during `vagrant up` run on Windows - but this could be a separate thing:

```
: "C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe" "C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log" 
: Log file location: F:\DevareaLocal\Prominic\GIT\VagrantRoyaleExample\vagrant_up.log 
: Bringing machine 'default' up with 'virtualbox' provider...
: ==> default: Clearing any previously set forwarded ports...
: ==> default: Clearing any previously set network interfaces...
: ==> default: Preparing network interfaces based on configuration...
:     default: Adapter 1: nat
: ==> default: Forwarding ports...
:     default: 8080 (guest) => 8086 (host) (adapter 1)
:     default: 22 (guest) => 2222 (host) (adapter 1)
: ==> default: Booting VM...
: C:\HashiCorp\Vagrant\bin\vagrant.exe : There was an error while executing `VBoxManage`, a CLI used by Vagrant
: At line:1 char:1
: + C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log
: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:     + CategoryInfo          : NotSpecified: (There was an er...used by Vagrant:String) [], RemoteException
:     + FullyQualifiedErrorId : NativeCommandError
:  
: for controlling VirtualBox. The command and stderr is shown below.
: Command: ["startvm", "5d3fc0a0-90d6-43e7-97ba-5b6f273767c4", "--type", "headless"]
: Stderr: VBoxManage.exe: error: The virtual machine 'VagrantRoyaleExample_default_1641467976898_91276' has terminated 
: unexpectedly during startup with exit code 1 (0x1).  More details may be available in 'C:\Users\Shadow\VirtualBox 
: VMs\VagrantRoyaleExample_default_1641467976898_91276\Logs\VBoxHardening.log'
: VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component MachineWrap, interface IMachine
```

**Comment by rat-moonshine** [comment]
On executing PowerShell through batch-command on Windows, I have adjusted Vagrant run-codes to accommodate PowerShell path to run with `tee`:
- 32/64 bit path detection 
- Running with `tee` instead of earlier `type` command - I confirm the outputs are available immediately to Moonshine console
- Shall notify in console if PowerShell is not available on any particular Windows

However, I see other problem during `vagrant up` run on Windows - but this could be a separate thing:

```
: "C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe" "C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log" 
: Log file location: F:\DevareaLocal\Prominic\GIT\VagrantRoyaleExample\vagrant_up.log 
: Bringing machine 'default' up with 'virtualbox' provider...
: ==> default: Clearing any previously set forwarded ports...
: ==> default: Clearing any previously set network interfaces...
: ==> default: Preparing network interfaces based on configuration...
:     default: Adapter 1: nat
: ==> default: Forwarding ports...
:     default: 8080 (guest) => 8086 (host) (adapter 1)
:     default: 22 (guest) => 2222 (host) (adapter 1)
: ==> default: Booting VM...
: C:\HashiCorp\Vagrant\bin\vagrant.exe : There was an error while executing `VBoxManage`, a CLI used by Vagrant
: At line:1 char:1
: + C:\HashiCorp\Vagrant\bin\vagrant.exe up 2>&1 | tee vagrant_up.log
: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:     + CategoryInfo          : NotSpecified: (There was an er...used by Vagrant:String) [], RemoteException
:     + FullyQualifiedErrorId : NativeCommandError
:  
: for controlling VirtualBox. The command and stderr is shown below.
: Command: ["startvm", "5d3fc0a0-90d6-43e7-97ba-5b6f273767c4", "--type", "headless"]
: Stderr: VBoxManage.exe: error: The virtual machine 'VagrantRoyaleExample_default_1641467976898_91276' has terminated 
: unexpectedly during startup with exit code 1 (0x1).  More details may be available in 'C:\Users\Devsena\VirtualBox 
: VMs\VagrantRoyaleExample_default_1641467976898_91276\Logs\VBoxHardening.log'
: VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component MachineWrap, interface IMachine
```

**Comment by rat-moonshine** [comment]
> I tried to open vagrant_up.log in Moonshine, and I got this prompt:
> 
> ![image](https://user-images.githubusercontent.com/24532731/150585223-902ddbf6-1bea-4871-a6e4-ee84a6c0f84b.png)
> 
> Moonshine should treat .log files as .txt files. Is this checking something besides the extension?

I see `.log` file opens in Windows Moonshine, but on macOS.

Moonshine tries to determine a file if binary by its content - here's the current code that to determine if a file is binary - https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/utils/UtilsCore.as#L150. I'm not sure what in the log file makes the difference to determine between Windows and macOS, but I'd prefer to adjust the logic rather than adding any explicit handling for `.log` file.

**Comment by rat-moonshine** [comment]
> Vagrant > SSH did not work. It opened a Command Line to the correct directory, but the vagrant ssh command didn't run.

This should be fixed now.

**Comment by JoelProminic** [comment]
> > I tried to open vagrant_up.log in Moonshine, and I got this prompt:
> > ![image](https://user-images.githubusercontent.com/24532731/150585223-902ddbf6-1bea-4871-a6e4-ee84a6c0f84b.png)
> > Moonshine should treat .log files as .txt files. Is this checking something besides the extension?
> 
> I see `.log` file opens in Windows Moonshine, but on macOS.
> 
> Moonshine tries to determine a file if binary by its content - here's the current code that to determine if a file is binary - https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/utils/UtilsCore.as#L150. I'm not sure what in the log file makes the difference to determine between Windows and macOS, but I'd prefer to adjust the logic rather than adding any explicit handling for `.log` file.

We saw some cases today where we could open a vagrant .log file in Moonshine.

Looking at the code snippet, I see that it considers files binary if it finds [unicode characters](https://en.wikipedia.org/wiki/List_of_Unicode_characters) in the ranges 00-08 or 0E-1F.   I did some searching in a failing file, and I found a 0x1B character (ESC).

It looks like vagrant is writing some control characters to change the position of the cursor, so we may find similar problems for other characters.  We may need to investigate a better solution for detecting binary files.

**Comment by JoelProminic** [comment]
VirtualBox Home seems to work like I expect now after @rat-moonshine fixed a bug we found in the morning chat.

**Comment by rat-moonshine** [comment]
> It looks like vagrant is writing some control characters to change the position of the cursor, so we may find similar problems for other characters. We may need to investigate a better solution for detecting binary files.

Looking through StackOverflow articles, it seems the most discussed and obvious solution on Unix is using `file`:
`file -b --mime-type fileName`

Unfortunately, a corresponding solution not easy available for Windows. There is a Win32 port of `file` [available](http://gnuwin32.sourceforge.net/packages/file.htm) but it seems to have many of its overheads to use, on Windows.

**Comment by rat-moonshine** [comment]
A temporary fix added to open problematic `.log` file for now. We shall remove the fix once #966 been addressed.

**Comment by JoelProminic** [comment]
I confirmed that I can open vagrant_up.log and other .log files now, at least on macOS Catalina.

There was a bug earlier today where Vagrantfile would not open with the below error:

    : Workspace changed to: Vagrant
    : Error #1009
    : TypeError: Error #1009
    : 	at Function/actionScripts.controllers:OpenFileCommand/protected:prepareBeforeOpen/actionScripts.controllers:fileLoadCompletes()

But @rat-moonshine has already resolved the problem.

**Comment by JoelProminic** [comment]
@Aszusz and @MarkProminic reported that if we run `Vagrant > Destroy` and then try `Vagrant > Up` again, then we get errors because some Ruby processes are still running.   I would like to investigate this before we close the issue, but otherwise this is looking good.

**Comment by rat-moonshine** [comment]
In a cleanly environment, I couldn't able to reproduce the Ruby process hold-up situation when tested between `destroy` and `up` respectively.

1. Chose `Vagrant > Destroy` from the context-menu
2. Console starts populating with output, a process/progress-bar shown up at bottom right corner
3. After some time the process over and the progress-bar disappeared from bottom right corner
4. Chose `Vagrant > Up`
5. A new process being started without any error 

I wonder if Ruby process hold-up is reason of running many different Vagrant command in a session and if also terminates some process abruptly.

**Comment by JoelProminic** [comment]
Ugh, I had done some tests for this last week, but I don't see this in a comment:

Test | Result
-----|-------
Destroy then Up | Works properly
Retry above | Works properly
Run up, then kill with red button | All processes closed
Run up, then Halt | `ruby` command remains open for a few seconds until `vagrant up` finishes

I remember the last case was a problem with the original VagrantRoyaleExample, since that script itentionally never finished.

I confirmed that killing the "vagrant up" command is not enough to stop the ruby process. 
1. Start Vagrant > Up in Moonshine. A fresh run will give you the more time to run this test
2. Kill the "vagrant up" process (the one with lots of environment variables)
3. Confirm that there is still a running `ruby` process
4. Try another `vagrant up` or Vagrant > Up.  The command fails with the below error:

    An action 'up' was attempted on the machine 'default',
    but another process is already executing an action on the machine.
    Vagrant locks each machine for access by only one process at a time.
    Please wait until the other Vagrant process finishes modifying this
    machine, then try again.

    If you believe this message is in error, please check the process
    listing for any "ruby" or "vagrant" processes and kill them. Then
    try again.

**Comment by JoelProminic** [comment]
I moved the hanging process issue to #970. 

Otherwise, I think basic Vagrant support is mostly working well.
